### DISCUSSION TITLE
Error while decrypting segment: Padding is incorrect

### DISCUSSION THREAD

**Discussion by zhagsd1** [streamlink/streamlink#4947] [Q&A] [answered]
### Checklist

- [X] This is a bug report and not a different kind of issue
- [X] [I have read the contribution guidelines](https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#contributing-to-streamlink)
- [X] [I have checked the list of open and recently closed bug reports](https://github.com/streamlink/streamlink/issues?q=is%3Aissue+label%3A%22bug%22)
- [X] [I have checked the commit log of the master branch](https://github.com/streamlink/streamlink/commits/master)

### Streamlink version

Latest build from the master branch

### Description

Hi First of all I have to say thank you. I spent the (whole) past 3 weeks just on this, from knowing abolutely nothing about coding and video streaming to little success step by step, even tho still a failure at this moment.

I attempted from yt-dlp to here. Thanks to eaiser key decryption integration with streamlink, and this post below really guides me through a lot of the thrill along the way [https://github.com/streamlink/streamlink/issues/2253](https://github.com/streamlink/streamlink/issues/2253), little sad its closed. I do have the same issues as everyone else at the very bottom of that post and was hoping to follow up on that. Also similar to [https://github.com/streamlink/streamlink/issues/4528](https://github.com/streamlink/streamlink/issues/4528) but with slightly different error wording and different video host server.

### Debug log

```text
>streamlink -l debug "https://d2qny97nu4rj64.cloud🦷front.net/spees/w/o/6197321b0cf23d7cc6e572e1/v/631cca34e4b0e9ba4c0511cf/u/621b96ad0cf202ac816dd59b/t/61d879beb5dc0add9bedfb56e929d434/p/assets/videos/6197321b0cf23d7cc6e572e1/2022/09/10/631cca34e4b0e9ba4c0511cf/hls_2M_.m3u8" best -o a1.mp4
NotVCSError: Git not installed; assuming this isn't a Git repository
[cli][debug] OS:         Windows 10
[cli][debug] Python:     3.10.6
[cli][debug] Streamlink: 0.0.0+unknown
[cli][debug] Dependencies:
[cli][debug]  isodate: 0.6.1
[cli][debug]  lxml: 4.9.1
[cli][debug]  pycountry: 22.3.5
[cli][debug]  pycryptodome: 3.15.0
[cli][debug]  PySocks: 1.7.1
[cli][debug]  requests: 2.28.1
[cli][debug]  websocket-client: 1.4.2
[cli][debug] Arguments:
[cli][debug]  stream=['best']
[cli][debug]  --loglevel=debug
[cli][debug]  --output=a1.mp4
[cli][info] Available streams: live (worst, best)
[cli][info] Opening stream: live (hls)
[cli][info] Writing output to
C:\Users\xxx\Downloads\Streamlink\a1.mp4
[cli][debug] Checking file output
[stream.hls][debug] Reloading playlist
[cli][debug] Pre-buffering 8192 bytes
[stream.hls][debug] Segments in this playlist are encrypted
[stream.hls][debug] First Sequence: 0; Last Sequence: 2475
[stream.hls][debug] Start offset: 0; Duration: None; Start Sequence: 0; End Sequence: 2475
[stream.hls][debug] Adding segment xx to queue
```

**Comment by bastimeyer** [comment, marked as answer]
The segments of this HLS playlist are encrypted and an AES-128 decryption key needs to be retrieved which needs to be used in addition to the initialization vector provided by the playlist, so that segment contents can be decrypted. This is regular stuff according to the HLS specification.

> `#EXT-X-KEY:METHOD=AES-128,URI="k/timestamp",IV=0x496daa1c6914000e408c65cead91fc29`

You will need to replace the value of the `URI` attribute with something meaningful. `k/timestamp` is not something Streamlink understands. Usually, and also according to the specification, the `URI` value needs to point to a location where the key can be retrieved from. This can either be an absolute or relative HTTP URL, or a custom URI with a custom scheme which the HLS implementation can understand. It can also be something entirely else, depending on the custom HLS implementation where this stream is meant to be played, like in this case.

If you're developing your own plugin for this, then you need to subclass the various HLSStream and M3U8Parser classes. Take a look at the Twitch or Abematv plugin for how to do this, but from taking a quick look at your plugin, you have already done that, albeit a bit awkwardly. Also not sure where you've got the old version of the `create_decryptor` method from.

My advice is, without having looked at anything specific, subclass the M3U8Parser, replace the `URI` attribute of the `EXT-X-KEY` HLS tag with a custom scheme and add a `requests` adapter hook for that scheme. Then do the necessary stuff in the hook for retrieving the AES decryption key.

Good luck.

**Reply by zhagsd1** [reply]
Thank you so so much.
I've been reading more of the code examples mentioned by you., which didn't mention the IV directly. 
I figure since the Error is Padding is incorrect, could it be directly linked to cloudfront.py I wrote Line 49?

```
        iv = key.iv or num_to_iv(sequence)
        # Pad IV if needed
        iv = b"\x00" * (16 - len(iv)) + iv
```

Any hints at how can i rewrite **b"\x00" * (16 - len(iv)) + iv** 😅 Thank you so much

**Reply by bastimeyer** [reply]
You are not supposed to copy outdated random pieces of code when using the latest version of Streamlink. The default [`create_decryptor`](https://github.com/streamlink/streamlink/blob/61794c1d28289ce44d7c74d9ccde054c8263c711/src/streamlink/stream/hls.py#L91-L127) method is fine as it is and you don't need to override it if you do what I said with adding the appropriate hooks for modifying the key data.

> Error while decrypting segment xx: PKCS#7 padding is incorrect.

This describes invalid data after decryption, not an invalid initialization vector. The IV is already embedded in the HLS playlist and can be used right away. The lines you've posted are completely irrelevant and are there as a fallback for incorrect IVs that violate the specification.

The error is coming from here:
https://github.com/streamlink/streamlink/blob/61794c1d28289ce44d7c74d9ccde054c8263c711/src/streamlink/stream/hls.py#L236-L250

Before data gets encrypted in block-encryption schemes like AES, it first needs to be encoded in a certain way, so that its length can be figured out again after decryption. AES-128 encrypts whole blocks of 16 bytes, which means that when you try to encrypt data of random length, you somehow need to know its length after decrypting it again without having any additional unencrypted metadata. The PKCS#7 padding algorithm ensures that the decrypted output data will have its correct length. It does that by adding padding zeroes and one byte for the data length to the data block. The unpadding simply reverts this.

You can use any arbitrary key to decrypt AES blocks. Whether the resulting data is correct and makes sense afterwards depends on the key you've used (and also the mode of operation of the decryption scheme, which is Cipher Block Chaining in case of HLS, but that's irrelevant here). This means that if the decrypted data is correct, then so will be the padding of the data and the unpadding will work just fine. However, if an incorrect decryption key was used and the decrypted data is just junk, then the unpad function that needs to be called afterwards can't make sense of this data unless the very very unlikely case happens that bits just align in the right way.

TL;DR your decryption key is most likely wrong.

**Reply by bastimeyer** [reply]
What you're asking for is already getting dangerously close to the limits of support that anybody here is willing to provide on this discussion forum. Nobody's going to implement a very specific custom plugin or key extraction logic for you. You'll have to figure this out on your own. Donating to the open collective doesn't change that :/

Maybe take a look at the `stringify` function. You're also ignoring the `convertArrayToBase64` calls for `g` and `m`. I can't give you an answer otherwise.

The HTTP response of the key request is also different every time while the IV and segment contents stay the same, which implies that there's more to it than this method which decrypts the first and last 16-byte-blocks of this 48-byte data.

**Reply by zhagsd1** [reply]
Sorry I was being extremely dumb. It turns out that my hands just too slow and the key updated already every single time (for all my attempts these three weeks). Your plugin works like a miracle! I think you can close #4528 too its a similar server so probably the same thing

**Reply by zhagsd1** [reply]
Hi just one last question? what should I do when the audio and video are two different url? one is .../hls_2M_.ts one is .../hls_audio_.ts Is there a command I can pass the audio and video url at the same time or I need to edit MuxedHLSStream()?

**Reply by bastimeyer** [reply]
> edit MuxedHLSStream

Why edit? Take a look at the parameters of `MuxedHLSStream`'s constructor and return an instance of it, just like a regular `HLSStream` instance. `MuxedHLSStream` requires a URL for a media playlist consisting of a main stream with video (and optional other data streams) and media playlist URLs with additional audio streams. If you have a multivariant playlist and you're parsing it via `HLSStream.parse_variant_playlist()`, then `MuxedHLSStream`s will be returned automatically depending on the multivariant playlist contents and the availability of ffmpeg on your system.

**Reply by zhagsd1** [reply]
Ah ok thank you. My .../index.m3u8 url only returns 3 video options and the audio is different, but I'll just download them seperatedly and mux them later then
